plutus-playground-lib/src/Schema.hs

```diff
@@ -0,0 +1,300 @@
+
+formArgumentToJson :: Fix FormArgumentF -> Maybe JSON.Value
+formArgumentToJson = cata algebra
+  where
+    algebra :: FormArgumentF (Maybe JSON.Value) -> Maybe JSON.Value
+    algebra FormUnitF = Just $ toJSON ()
+    algebra (FormBoolF v) = Just $ toJSON v
+    algebra (FormIntF v) = Just $ toJSON v
+    algebra (FormStringF v) = Just $ toJSON v
+    algebra (FormHexF v) = Just $ toJSON v
+    algebra (FormRadioF _ v) = Just $ toJSON v
+    algebra (FormArrayF _ v) = Just $ toJSON v
+    algebra (FormMaybeF _ v) = Just $ toJSON v
+    algebra (FormTupleF (Just a) (Just b)) = Just $ toJSON [a, b]
+    algebra (FormTupleF _ _) = Nothing
+    algebra (FormObjectF vs) =
+        JSON.Object . HashMap.fromList . map (first Text.pack) <$>
+        traverse sequence vs
+    algebra (FormValueF v) = Just $ toJSON v
+    algebra (FormSlotRangeF v) = Just $ toJSON v
+    algebra (FormUnsupportedF _) = Nothing
+
+------------------------------------------------------------
+class ToSchema a where
+    toSchema :: FormSchema
+    default toSchema :: (Generic a, GenericToSchema (Rep a)) =>
+        FormSchema
+    toSchema = genericToSchema $ from (undefined :: a)
+
+class ToArgument a where
+    toArgument :: a -> Fix FormArgumentF
+    default toArgument :: (Generic a, GenericToArgument (Rep a)) =>
+        a -> Fix FormArgumentF
+    toArgument = genericToArgument . from
+
+instance ToSchema () where
+    toSchema = FormSchemaUnit
+
+instance ToArgument () where
+    toArgument _ = Fix FormUnitF
+
+instance ToSchema Bool where
+    toSchema = FormSchemaBool
+
+instance ToArgument Bool where
+    toArgument = Fix . FormBoolF
+
+instance ToSchema Int where
+    toSchema = FormSchemaInt
+
+instance ToArgument Int where
+    toArgument = Fix . FormIntF . Just
+
+instance ToSchema Integer where
+    toSchema = FormSchemaInt
+
+instance ToArgument Integer where
+    toArgument = Fix . FormIntF . Just . fromIntegral
+
+instance ToSchema Text where
+    toSchema = FormSchemaString
+
+instance ToArgument Text where
+    toArgument = Fix . FormStringF . Just . Text.unpack
+
+instance ToSchema a => ToSchema (Proxy a) where
+    toSchema = toSchema @a
+
+instance (ToSchema k, ToSchema v) => ToSchema (Data.Map.Map k v) where
+    toSchema = FormSchemaArray $ toSchema @(k, v)
+
+instance (ToSchema a, ToSchema b) => ToSchema (a, b) where
+    toSchema = FormSchemaTuple (toSchema @a) (toSchema @b)
+
+instance (ToArgument a, ToArgument b) => ToArgument (a, b) where
+    toArgument (a, b) = Fix $ FormTupleF (toArgument a) (toArgument b)
+
+instance ToSchema String where
+    toSchema = FormSchemaString
+
+instance {-# OVERLAPPABLE #-} (ToSchema a) => ToSchema [a] where
+    toSchema = FormSchemaArray (toSchema @a)
+
+instance ToArgument String where
+    toArgument = Fix . FormStringF . Just
+
+instance {-# OVERLAPPABLE #-} (ToSchema a, ToArgument a) => ToArgument [a] where
+    toArgument xs = Fix $ FormArrayF (toSchema @a) (toArgument <$> xs)
```

Comment:
what do these overlap with?

Reply:
This is to deal with the fact that `String` is a `[Char]`.

Reply:
gah, haskell strings strike again!

Reply:
![Skeletor](http://www.riffsy.com/image/fef28cf0d58d3559a35f79223ea4a7d9-2.gif)